### DISCUSSION TITLE
Allow casting to open generic type to access non-generic members

### DISCUSSION THREAD

**Discussion by TheXenocide** [dotnet/csharplang#7108] [Language Ideas]
Prototyping a new fun project recently I realized that there's a pattern I've seen often that could possibly be simplified with some help from a smarter compiler. The thought is that open generic type definitions still have plenty of non-generic contract definitions and it would be nice to be able to use them. It seems like there could be a reason this hasn't been done, but in case it was simply excluded because it wasn't thought to be useful, imagine the following example:

```cs
public abstract class NameValueElementBase : Element
{
    public int Id { get; set; }
    public string Name {get; set; }
    
    public abstract Type ValueType { get; }
    public abstract object GetValue();
}

public class NameValueElement<TValue> : NameValueElementBase
{
    public TVaue Value { get; set; }
    public override Type ValueType => typeof(TValue);
    public override object GetValue() => (object)Value;
}

public class SampleUsage
{
    public void VisitElement(Element elements)
    {
        if (element is NamedValueElementBase nv) {
            DoSomething(nv.Name);
        }
        // ...
    }
}
```

In this scenario the main reason for the existence of the base class is to allow casting to access non-generic contract definitions which, in turn allows many different generic types based on this definition to be included in generic collections and so on. On the flip-side, this type needs to be as publicly accessible as the generic type, even if it isn't intended to be derived from by anything but the generic type it supports; it's contract filler (perhaps a little less necessary considering interfaces can now include default implementations). It seems like this could be simplified by allowing the use of an open generic as a contract against its non-generic members. Is there a particularly strong reason why the following example shouldn't be supported by the language?

```cs
public class NameValueElement<TValue> : Element
{
    public int Id { get; set; }
    public string Name {get; set; }

    public TVaue Value { get; set; }
    
    public Type ValueType => typeof(TValue);
    public object GetValue() => (object)Value;
}

public class SampleUsage
{
    public void VisitElement(Element elements)
    {
        if (element is NamedValueElement<> nv) {
            DoSomething(nv.Name);
        }
        // ...
    }
}
```

It also seems like it could be helpful to apply this in collections with covariance, e.g. `IEnumerable<NameValueElement<>>` could be assigned from `List<NameValueElement<string>>`.

**Comment by HaloFour** [comment]
This would require runtime support as currently the runtime forbids using open generic types like this.  I'd love to see something like Java's bounded generic wildcards, but I imagine that it would involve a massive amount of work.

See also: #1992

**Reply by TheXenocide** [reply]
It has been a long time since I used Java generics, but if I recall the way they implement generics does not lend as well to the types of optimizations possible with .NET's and I'd rather keep the fundamentals more like .NET, but this suggestion doesn't seem like a far stretch from the existing APIs at least. I understand this would probably involve introducing new IL, JIT optimizations, etc. so I'm not saying it would be easy, just feels like it would be useful and I haven't thought of any reasons why it wouldn't be statically verifiable/JIT compilable/etc. within the existing language infrastructure. What this proposes perhaps keeps an initial implementation more simple but progresses well toward implementing a lower level/more optimized, but similarly flexible language feature/extension like the approach mentioned in #1992. 

Then again, from recent dabbling I could also stand for something as "simple" as a generic type constraint that can infer the base type of an enum into a generic parameter, which suddenly starts to sound an awful like something that would be possible with what's suggested over there. My initial experiments to make a DB type model wanted something like `class EnumEntityBase<T, TBase> where T : Enum, TBase` or some such to allow using the underlying primitive type for the `Id` property all nice and statically checked. Instead, even if I know the underlying type explicitly (e.g. `Int32EnumEntityBase<T>`) I have to box and unbox the value to cast (e.g. `return (int)(object)tTypedValue`). Seems like a `where T : Enum<?TBase>` is something like the syntax being proposed over there might be a decent solution for this scenario.

**Reply by HaloFour** [reply]
Java's generics are erased at runtime, which does make them less capable but also allows them to be a lot more flexible.  The runtime doesn't enforce anything, so the language is free to enforce variance as it sees fit.  The exact behavior you described is a normal feature of Java generics:

```java
int getListLength(List<?> list) {
    return list.length();
}
```

.NET's generics are what makes this much harder, it's not really possible without runtime support.  This is why it's pretty typical in this case to declare a non-generic base interface which defines the members that don't need to use the generic type.

**Reply by TheXenocide** [reply]
I suppose there is also the option of the compiler automatically generating the non-generic base type and interpreting/compiling references to the open(-looking) declaration to use the generated base type, though I'm actually somewhat partial to the runtime-feature-supported approach it does seem rather involved.
